### PR TITLE
add --light-intensity option

### PR DESCRIPTION
**Pull request by snoyer**
This PR adds a `--light-intensity` option to address problem described in #352.

It is implemented by simply adding a [`vtkLightKit`](https://vtk.org/doc/release/4.0/html/classvtkLightKit.html) to the renderer.

Here are some samples using the same model as in previous issue:

```bash
f3d 'ROKN Chung Mu.glb' --camera-position 37.3,20.4,61.6 --camera-focal-point -5.5,1.4,6.7 --camera-view-up -0.2,1.0,-0.2 --camera-view-angle 30 --ssao --fxaa --tone-mapping
```
light-intensity-no

```bash
f3d 'ROKN Chung Mu.glb' --light-intensity=1 --camera-position 37.3,20.4,61.6 --camera-focal-point -5.5,1.4,6.7 --camera-view-up -0.2,1.0,-0.2 --camera-view-angle 30 --ssao --fxaa --tone-mapping --light-intensity=1
```
light-intensity-1

```bash
f3d 'ROKN Chung Mu.glb' --light-intensity=2 --camera-position 37.3,20.4,61.6 --camera-focal-point -5.5,1.4,6.7 --camera-view-up -0.2,1.0,-0.2 --camera-view-angle 30 --ssao --fxaa --tone-mapping --light-intensity=2
```
light-intensity-2

```bash
f3d 'ROKN Chung Mu.glb' --light-intensity=4 --camera-position 37.3,20.4,61.6 --camera-focal-point -5.5,1.4,6.7 --camera-view-up -0.2,1.0,-0.2 --camera-view-angle 30 --ssao --fxaa --tone-mapping --light-intensity=4
```
light-intensity-4

### CODE REVIEW

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:535]: What happens if you set the lightIntensity to 0 instead ?

File: application/F3DOptionsParser.cxx
Comment (mwestphal) [inline application/F3DOptionsParser.cxx:249]: README.md update needed

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.h:180]: Isn't there already a default light kit in all scenes ? What about Renderer->GetLights() ?

**Comment by mwestphal** [comment]
Hi @snoyer 

Thanks a lot for your contribution. This is a needed feature, but we will need to make sure this is the right implementation :). Let's wait for @Meakk feedback on this.

In the meantime, you may want to add tests.

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (snoyer) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.h:180]: According to the vtkLightKit doc

> To enforce the relationship between the three lights, the intensity of the fill and headlights are set as a ratio to the key light brightness. Thus, the brightness of all the lights in the scene can be changed by changing the key light intensity.

so having a handle on a full light kit rather than individual lights would make changing the intensity consistently easier I would guess? (disclaimer: again, I know nothing about f3d or VTK)

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx
Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:535]: Ok, lets see what @Meakk thinks of that.

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.h:180]: tbh @snoyer , for a F3D and VTK beginner, this is great code !

We will check that and get back to you.

**Review by Meakk — CHANGES_REQUESTED** [review]
Thanks for this @snoyer!  
I think the current approach might cause issues in some cases.  
Let me explain quickly how lights are handled in F3D.  

There are two code paths depending if the file format support "full scene" mode (e.g. OBJ, GLTF), or "geometry only" (if file format does not support lights like STL for example, or if `--geometry-only` option is set).  
If the mode is "geometry only", then we are using a vtkLightKit (see `vtkF3DGenericImporter::ImportLights`)  
If the move is "full scene", then the VTK importer is responsible of adding the lights (most of the time a single headlight is added, but we can have more complicated light configuration)  
So adding a `vtkLightKit` if `--light-intensity` is set will not change the intensity of the lights added by the importer.  

My recommendation is to loop on existing lights (Use `this->GetLights()`) and change all the intensities using `vtkLight::SetIntensity` in your function `vtkF3DRenderer::SetLightIntensity`.  
Don't forget to call this function everytime by adding it in `window_impl::UpdateDynamicOptions` to make sure it's properly called when a new file is loaded.

**Comment by Meakk** [comment]
For the loop, something like that should work:

```cpp
  vtkLight* light;
  vtkLightCollection* lc = this->GetLights();
  vtkCollectionSimpleIterator it;
  for (lc->InitTraversal(it); (light = lc->GetNextLight(it));)
  {
    light->SetIntensity(intensity);
  }
```

**Comment by snoyer** [comment]
> Thanks for this @snoyer! I think the current approach might cause issues in some cases. Let me explain quickly how lights are handled in F3D.
> 
> There are two code paths depending if the file format support "full scene" mode (e.g. OBJ, GLTF), or "geometry only" (if file format does not support lights like STL for example, or if `--geometry-only` option is set). If the mode is "geometry only", then we are using a vtkLightKit (see `vtkF3DGenericImporter::ImportLights`) If the move is "full scene", then the VTK importer is responsible of adding the lights (most of the time a single headlight is added, but we can have more complicated light configuration) So adding a `vtkLightKit` if `--light-intensity` is set will not change the intensity of the lights added by the importer.

Thanks @Meakk, this is very helpful.

> My recommendation is to loop on existing lights (Use `this->GetLights()`) and change all the intensities using `vtkLight::SetIntensity` in your function `vtkF3DRenderer::SetLightIntensity`. Don't forget to call this function everytime by adding it in `window_impl::UpdateDynamicOptions` to make sure it's properly called when a new file is loaded.

Alright, @mwestphal also suggested working with existing lights and that all makes sense, I have adjusted the code accordingly.

**Comment by snoyer** [comment]
Interestingly, this works in window mode but not always in mode.

It looks like `window_impl::UpdateDynamicOptions` is called both before and after the lights get added when in window mode, but when `--output` is set it is only called before and not after.

**Comment by mwestphal** [comment]
@snoyer This sounds like an actual bug, I will take a look.

**Comment by mwestphal** [comment]
I reproduce the issue, I'll take a look.

**Comment by mwestphal** [comment]
@Meakk Turns out there is a **third** light code path

```
  ///@{
  /**  
   * Turn on/off a flag which disables the automatic light creation capability.
   * Normally in VTK if no lights are associated with the renderer, then a light
   * is automatically created. However, in special circumstances this feature is
   * undesirable, so the following boolean is provided to disable automatic
   * light creation. (Turn AutomaticLightCreation off if you do not want lights
   * to be created.)
   */
  vtkGetMacro(AutomaticLightCreation, vtkTypeBool);
  vtkSetMacro(AutomaticLightCreation, vtkTypeBool);
  vtkBooleanMacro(AutomaticLightCreation, vtkTypeBool);
  ///@}

```


`vtkOpenGLRenderer::UpdateLights()`
```
  // create alight if needed
  if (!lightingCount)
  {
    if (this->AutomaticLightCreation)
    {    
      vtkDebugMacro(<< "No lights are on, creating one.");
      this->CreateLight();
      lc->InitTraversal(sit);
```

When an importer do not provide any light, VTK kindly add a single point light. However it adds it only one the first render, and this is too late for @snoyer feature.

I will investigate how to fix that.

**Comment by mwestphal** [comment]
@snoyer : here is the fix: https://github.com/f3d-app/f3d/pull/463

You can either rebase on my branch or just copy the lines, as you see fit.

**Comment by mwestphal** [comment]
@snoyer : I would suggest a rebase instead of a merge, cleaner git history.

**Comment by snoyer** [comment]
> @snoyer : I would suggest a rebase instead of a merge, cleaner git history.

Sorry, I suck at git and did the lazy thing of updating the branch form the website :/

**Comment by mwestphal** [comment]
No worries, I'll check that before the merge :)

File: application/F3DOptionsParser.h
Comment (snoyer) [inline application/F3DOptionsParser.h:43]: this is actually unused, should be removed

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.h:181]: member var start with capital letter: OriginalLightIntensities

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:535]: I think this map should be cleaned on initialization

Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:535]: already done,

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:536]: how about:

```
double originalIntensity;
if (this->originalLightIntensities.contains(light))
{
  originalIntensity = this->originalLightIntensities[light];
}
else
{
  originalIntensity = light->GetIntensity();
  this->originalLightIntensities[light] = originalIntensity;
}
```

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.h:15]: vtkLight

**Review by mwestphal — CHANGES_REQUESTED** [review]
Still needs a test :)

File: library/VTKExtensions/Rendering/vtkF3DRenderer.cxx
Comment (snoyer) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.cxx:535]: `.clear()` in `vtkF3DRenderer::Initialize` not good?

File: library/testing/TestSDKDynamicLightIntensity.cxx
Comment (Meakk) [inline library/testing/TestSDKDynamicLightIntensity.cxx:8]: You could have added one line in `cmake/testing.cmake` it would have been easier, but it's fine the way you did it.  
@mwestphal we should add a section in the documentation explaining how to make test.

**Review by Meakk — COMMENTED** [review]
Please add the option in the documentation and it's good to go in my opinion.  
Thanks again for your contribution.

File: library/testing/TestSDKDynamicLightIntensity.cxx
Comment (snoyer) [reply library/testing/TestSDKDynamicLightIntensity.cxx:8]: Yeah I couldn't figure it out so I went with TestSDKDynamicBackgroundColor as a model.

In the end it worked out best this way because the test covers changing the intensity dynamically (ie. the test will fail if `this->OriginalLightIntensities[light] = originalIntensity;` is removed, which would not be a problem when `vtkF3DRenderer::SetLightIntensity` is only called once)

Comment (mwestphal) [reply library/testing/TestSDKDynamicLightIntensity.cxx:8]: Very good point @Meakk !

It would be nice to also add a --light-intensity test in cmake/testing.cmake though.

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [inline library/VTKExtensions/Rendering/vtkF3DRenderer.h:14]: std include should be put after vtk includes, with an empty lines in betweem

Comment (snoyer) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.h]: Could the formatter enforce that? something like:
```yaml
IncludeBlocks: Regroup
IncludeCategories:
  - Regex:           '".*"'
    Priority:        1
  - Regex:           '^<(vtk.*)>'
    Priority:        2
    CaseSensitive:   true
  # ... anything else ?
```

File: library/testing/TestSDKDynamicLightIntensity.cxx
Comment (snoyer) [reply library/testing/TestSDKDynamicLightIntensity.cxx:8]: I barely managed to write and run this one test by poking around and making educated guesses, I still don't know how to properly run the whole test suite.
You guys need to document how to 1. run the existing tests, and 2. write new ones if you don't want new contributors to give up before their feature is considered good enough

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.h]: good point, wdyt @Meakk ?

File: library/testing/TestSDKDynamicLightIntensity.cxx
Comment (mwestphal) [reply library/testing/TestSDKDynamicLightIntensity.cxx:8]: Ok, that's fair. We are currently rewriting the whole documentation but it is taking some time.

Thanks a lot for your patience and tenacity. I'll add another test once we merge this.

**Review by mwestphal — APPROVED** [review]

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.h]: alternative is to write a guide style, but such a doc are hard to write and maintain

File: library/testing/TestSDKDynamicLightIntensity.cxx
Comment (mwestphal) [reply library/testing/TestSDKDynamicLightIntensity.cxx:8]: Here is the doc, still a WIP, but should make things clearer:

https://github.com/mwestphal/f3d/blob/jekyll_doc/doc/dev/TESTING.md

**Review by Meakk — APPROVED** [review]

File: library/VTKExtensions/Rendering/vtkF3DRenderer.h
Comment (mwestphal) [reply library/VTKExtensions/Rendering/vtkF3DRenderer.h]: First try at a coding style doc: https://github.com/mwestphal/f3d/blob/jekyll_doc/doc/dev/CODING_STYLE.md

**Pull request merged**
